### PR TITLE
Resolve lifetimes independently for each item-like.

### PR DESCRIPTION
**Pull request by cjgillot**
Now that the heavy-lifting is done on the AST and during lowering, we do not need to perform HIR lifetime resolution on a full item at once.  Instead, we can treat each item-like independently, and look at `generics_of` the parent exceptionally for associated items.

### CODE REVIEW

File: src/test/ui/impl-trait/nested-rpit-hrtb.stderr [hunk 1]
```diff
@@ -65,18 +65,70 @@ note: lifetime declared here
...
+   = help: the trait `Qux<'_>` is implemented for `()`
+
+error: implementation of `Bar` is not general enough
+  --> $DIR/nested-rpit-hrtb.rs:51:93
+   |
+LL | fn one_hrtb_mention_fn_outlives_uses<'b>() -> impl for<'a> Bar<'a, Assoc = impl Sized + 'b> {}
+   |                                                                                             ^^ implementation of `Bar` is not general enough
+   |
+   = note: `()` must implement `Bar<'0>`, for any lifetime `'0`...
+   = note: ...but it actually implements `Bar<'1>`, for some specific lifetime `'1`
```
Comment (estebank) [inline src/test/ui/impl-trait/nested-rpit-hrtb.stderr:112]: Can we do something to point at the return type here? Maybe special case functions that have no tail expression nor return statement?

File: src/test/ui/impl-trait/impl-fn-hrtb-bounds.stderr [hunk 2]
```diff
@@ -22,30 +22,57 @@ note: lifetime declared here
...
    |                   ^
 
+error: lifetime may not live long enough
+  --> $DIR/impl-fn-hrtb-bounds.rs:6:9
+   |
+LL |     |x| x
+   |      -- ^ returning this value requires that `'1` must outlive `'2`
+   |      ||
+   |      |return type of closure is impl Debug + '2
+   |      has type `&'1 u8`
```
Comment (estebank) [inline src/test/ui/impl-trait/impl-fn-hrtb-bounds.stderr:32]: Could we somehow silence this? (Can be follow up work, but do we have a clear way to do it?)

File: src/test/ui/impl-trait/nested-rpit-hrtb.rs [hunk 3]
```diff
@@ -31,34 +31,40 @@ fn one_hrtb_trait_param() -> impl for<'a> Foo<'a, Assoc = impl Qux<'a>> {}
...
 fn two_htrb_trait_param() -> impl for<'a> Foo<'a, Assoc = impl for<'b> Qux<'b>> {}
 
 // `'b` is not in scope for the outlives bound.
 fn two_htrb_outlives() -> impl for<'a> Foo<'a, Assoc = impl for<'b> Sized + 'b> {}
 //~^ ERROR use of undeclared lifetime name `'b` [E0261]
 
-// This should pass.
+// This should resolve.
 fn two_htrb_trait_param_uses() -> impl for<'a> Bar<'a, Assoc = impl for<'b> Qux<'b>> {}
+//~^ ERROR the trait bound `for<'b> &(): Qux<'b>` is not satisfied
```
Comment (estebank) [inline src/test/ui/impl-trait/nested-rpit-hrtb.rs:63]: Wait, isn't this a real regression?

File: compiler/rustc_hir_analysis/src/collect/lifetimes.rs [hunk 4]
```diff
@@ -766,65 +689,29 @@ impl<'a, 'tcx> Visitor<'tcx> for LifetimeContext<'a, 'tcx> {
...
-                                self.tcx.sess,
-                                lifetime.span,
-                                E0657,
-                                "`impl Trait` can only capture lifetimes \
+                        struct_span_err!(
+                            self.tcx.sess,
+                            lifetime.span,
+                            E0657,
+                            "`impl Trait` can only capture lifetimes \
                                     bound at the fn or impl level"
```
Comment (estebank) [inline compiler/rustc_hir_analysis/src/collect/lifetimes.rs:697]: We can clean up the formatting here.

**Review by estebank — APPROVED** [review]
The code changes overall look fine, but I had a bunch of questions.

File: src/test/ui/impl-trait/nested-rpit-hrtb.rs (hunk 3 above)
Comment (cjgillot) [reply src/test/ui/impl-trait/nested-rpit-hrtb.rs]: The error is legitimate. `Bar<'a>::Assoc` is `&'a ()`, which does not implement `Qux<'b>`.
For some reason, the compiler was aborting before before reaching this point.

File: src/test/ui/impl-trait/impl-fn-hrtb-bounds.stderr (hunk 2 above)
Comment (cjgillot) [reply src/test/ui/impl-trait/impl-fn-hrtb-bounds.stderr]: I don't really know why the compiler does not abort any more. I speculate that the `higher kinded lifetime bounds on nested opaque types are not supported yet` error is not emitted from the same code path any more, so the `track_errors` calls in `rustc_hir_analysis::check_crate` do not trigger any more.

Comment (cjgillot) [reply src/test/ui/impl-trait/impl-fn-hrtb-bounds.stderr]: I added a commit, which should remove some spurious errors.

**Pull request merged**